src/platform/windows/ipc/misc_utils.cpp

```diff
@@ -0,0 +1,293 @@
+    if (0xFFFFFFFF == consoleSessionId) {
+      return nullptr;
+    }
+
+    // Get the user token for the active console session
+    if (!WTSQueryUserToken(consoleSessionId, &userToken)) {
+      return nullptr;
+    }
+
+    // We need to know if this is an elevated token or not.
+    // Get the elevation type of the user token
+    if (!GetTokenInformation(userToken, TokenElevationType, &elevationType, sizeof(TOKEN_ELEVATION_TYPE), &dwSize)) {
+      CloseHandle(userToken);
+      return nullptr;
+    }
+
+    // User is currently not an administrator
+    if (elevated && (elevationType == TokenElevationTypeDefault && !IsUserAdmin(userToken))) {
+      // Don't elevate, just return the token as is
+    }
+
+    // User has a limited token, this means they have UAC enabled and is an Administrator
+    if (elevated && elevationType == TokenElevationTypeLimited) {
+      TOKEN_LINKED_TOKEN linkedToken;
+      // Retrieve the administrator token that is linked to the limited token
+      if (!GetTokenInformation(userToken, TokenLinkedToken, static_cast<void *>(&linkedToken), sizeof(TOKEN_LINKED_TOKEN), &dwSize)) {
+        CloseHandle(userToken);
+        return nullptr;
+      }
+
+      // Since we need the elevated token, we'll replace it with their administrative token.
+      CloseHandle(userToken);
+      userToken = linkedToken.LinkedToken;
+    }
+
+    // We don't need to do anything for TokenElevationTypeFull users here, because they're already elevated.
+    return userToken;
+  }
+
+  // Function to check if a process with the given name is running
+  bool is_process_running(const std::wstring &processName) {
+    HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
+    if (snapshot == INVALID_HANDLE_VALUE) {
+      return false;
+    }
+
+    PROCESSENTRY32W processEntry = {};
+    processEntry.dwSize = sizeof(processEntry);
+
+    bool found = false;
+    if (Process32FirstW(snapshot, &processEntry)) {
+      do {
+        if (_wcsicmp(processEntry.szExeFile, processName.c_str()) == 0) {
+          found = true;
+          break;
+        }
+      } while (Process32NextW(snapshot, &processEntry));
+    }
+
+    CloseHandle(snapshot);
+    return found;
+  }
+
+  // Function to check if we're on the secure desktop
+  bool is_secure_desktop_active() {
+    // Check for UAC (consent.exe)
+    if (is_process_running(L"consent.exe")) {
+      return true;
+    }
+
+    // Check for login screen by looking for winlogon.exe with specific conditions
+    // or check the current desktop name
+    if (HDESK currentDesktop = GetThreadDesktop(GetCurrentThreadId())) {
+      std::wstring desktopName(256, L'\0');
+      DWORD needed = 0;
+      if (GetUserObjectInformationW(currentDesktop, UOI_NAME, &desktopName[0], static_cast<DWORD>(desktopName.size() * sizeof(wchar_t)), &needed) && (_wcsicmp(desktopName.c_str(), L"Winlogon") == 0 || _wcsicmp(desktopName.c_str(), L"SAD") == 0)) {
```

Comment:
I wonder if we could do better here by perhaps using `GetUserObjectInformationW()` with `UOI_USER_SID` and checking for a specific SID (maybe `S-1-5-18` or the user's SID).